src/storage-controller/src/collection_mgmt.rs

```diff
@@ -446,16 +447,26 @@ where
         let handle = mz_ore::task::spawn(
             || format!("CollectionManager-incremental_write_task-{id}"),
             async move {
-                let _ = task.run().await;
+                let res = task.run().await;
 
-                info!("write_task-{} ending", id);
+                match res {
+                    ControlFlow::Break(reason) => {
+                        info!("write_task-{} ending: {}", id, reason);
+                    }
+                    c => {
+                        unreachable!(
+                            "cannot break out of the loop with a Continue, but got: {:?}",
+                            c
+                        );
+                    }
+                }
             },
         );
 
         (tx, handle.abort_on_drop(), shutdown_tx)
     }
 
-    async fn run(mut self) -> Result<(), ()> {
+    async fn run(mut self) -> ControlFlow<String> {
```

Comment:
Nice, didn't know this existed! Makes things much clearer.

Reply:
Yeah, I only remembered it because persist uses it quite a bit: https://github.com/MaterializeInc/materialize/blob/afc85d2748e3489536c9ebc051b87fa96bf59b6c/src/persist-client/src/internal/state.rs#L956 😅